haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs

```diff
@@ -137,16 +138,55 @@ analyseTypechecked' ghcEnv opts modSummary renSource tcSource (mod', mspan) = do
           renamedRefs = fromMaybe [] (refsFromRenamed ctx altMap <$> renSource)
       rels = fromMaybe [] (relationsFromRenamed ctx altMap <$> renSource)
       decls = map daDecl declsAlts
+      docDecls = generateDocDecls refs
   imports <- fromMaybe (return []) (importsFromRenamed ctx <$> renSource)
-  return $ XRef (AnalysedFile (SourcePath modFile) strippedModFile) moduleTick
-            decls refs rels imports
+  return $ XRef
+      { xrefFile = (AnalysedFile (SourcePath modFile) strippedModFile)
+      , xrefModule = moduleTick
+      , xrefDecls = decls
+      , xrefDocDecls = docDecls
+      , xrefCrossRefs = refs
+      , xrefRelations = rels
+      , xrefImports = imports
+      }
   where
     declAltMap :: [DeclAndAlt] -> DeclAltMap
     declAltMap = M.fromList . mapMaybe toPair
       where
         toPair (DeclAndAlt d (Just alt)) = Just (alt, declTick d)
         toPair _ = Nothing
 
+-- | Generate decls that point to hackage documents.
+-- These decls are mainly for things from GHC core packages. Their source code
+-- is not available but we can make them point to their documents.
+generateDocDecls :: [TickReference] -> [Decl]
+generateDocDecls refs =
+  -- Deduplicate by URI strings.
+  let uriTickMap =
+        foldr
+          ( \r acc ->
+              let targetTick = refTargetTick r
+               in case tickDocUri targetTick of
```

Comment:
Would it be more readable to write this as a list comprehension instead?
```
Map.fromList [(uri, targetTick) | r <- refs, let targetTick = refTargetTick r, Just uri <- tickDocUri targetTick]
```

Reply:
Agreed. I had to create a dummy list of `tickDocUri targetTick` though.